From 4e988561145e793ca45bd9a57d6b853485ed6dff Mon Sep 17 00:00:00 2001
From: Tamizh Chelvam Raja <quic_tamizhr@quicinc.com>
Date: Wed, 4 Oct 2023 22:49:26 +0530
Subject: [PATCH] wifi: ath12k: Fix tx and rx interrupt mask init and handling

Tx ring mask for QCN6432 used wrongly as ATH12K_TX_RING_MASK_3,
fix it by using correct mask ATH12K_TX_RING_MASK_4. And rx ring
mask is being shared between ring 3 and ring 4 but the interrupt
handling does not handle multiple masks in a single rx ring mask
group and that causes not handling interrupt for one ring.
Fix it by handling multiple ring masks in a single group.

Signed-off-by: Tamizh Chelvam Raja <quic_tamizhr@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/dp.c | 3 ++-
 drivers/net/wireless/ath/ath12k/hw.c | 2 +-
 2 files changed, 3 insertions(+), 2 deletions(-)

diff --git a/drivers/net/wireless/ath/ath12k/dp.c b/drivers/net/wireless/ath/ath12k/dp.c
index d571708..57061da 100644
--- a/drivers/net/wireless/ath/ath12k/dp.c
+++ b/drivers/net/wireless/ath/ath12k/dp.c
@@ -1345,8 +1345,9 @@ int ath12k_dp_service_srng(struct ath12k_base *ab,
 			goto done;
 	}
 
-	if (rx_mask) {
+	while (rx_mask) {
 		i =  fls(rx_mask) - 1;
+		rx_mask ^=  1 << i;
 		work_done = ath12k_dp_rx_process(ab, i, napi,
 						 budget);
 		budget -= work_done;
diff --git a/drivers/net/wireless/ath/ath12k/hw.c b/drivers/net/wireless/ath/ath12k/hw.c
index 5a3f6a5..4112112 100644
--- a/drivers/net/wireless/ath/ath12k/hw.c
+++ b/drivers/net/wireless/ath/ath12k/hw.c
@@ -949,7 +949,7 @@ static struct ath12k_hw_ring_mask ath12k_hw_ring_mask_qcn6432 = {
 	.tx  = {
 		ATH12K_TX_RING_MASK_0,
 		ATH12K_TX_RING_MASK_1,
-		ATH12K_TX_RING_MASK_2 |ATH12K_TX_RING_MASK_3,
+		ATH12K_TX_RING_MASK_2 | ATH12K_TX_RING_MASK_4,
 		0, 0, 0, 0, 0, 0, 0, 0,
 	},
 	.rx_mon_dest = {
-- 
2.17.1

